### PR TITLE
[#3315] Record duration up until error instead of hard coded 1 when using Metrics middleware

### LINKED ISSUE #3315: org.http4s.client.middleware.Metrics does not record elapsed time on timeout
The documentation/and comments on org.http4s.client.middleware.Metrics says that "Time duration of errors, timeouts and other abnormal terminations" is recorded.

But we see on line 73, that the duration is not recorded, but rather a hard coded value of 1 nano second is recorded.

https://github.com/http4s/http4s/blob/2c21e66d8312973ba159c6c271bf9e6225077b60/client/src/main/scala/org/http4s/client/middleware/Metrics.scala#L73

### PR DESCRIPTION
**Pull request by tian000**
This fixes #3315.

### CODE REVIEW

**Comment by tian000** [comment]
@rossabaker please take a look!

File: client/src/main/scala/org/http4s/client/middleware/Metrics.scala [hunk 1]
```diff
@@ -44,35 +44,38 @@ object Metrics {
...
-              oStatus.traverse_(status =>
-                ops.recordTotalTime(req.method, status, now - start, classifierF(req)))))
-      }
-      resp <- client.run(req)
-      _ <- Resource.liftF(statusRef.set(Some(resp.status)))
-      end <- Resource.liftF(clock.monotonic(TimeUnit.NANOSECONDS))
-      _ <- Resource.liftF(ops.recordHeadersTime(req.method, end - start, classifierF(req)))
-    } yield resp)
-      .handleErrorWith { (e: Throwable) =>
+      resp <- (for {
```
Comment (hamnis) [inline client/src/main/scala/org/http4s/client/middleware/Metrics.scala:50]: maybe put this in another method to improve readability

Comment (tian000) [reply client/src/main/scala/org/http4s/client/middleware/Metrics.scala]: @hamnis I broke the nesting out into functions. What do you think?

**Comment by tian000** [comment]
@hamnis The test that is failing seems related to `org.http4s.client.PoolManagerSpec`
How do I re-run the test?

**Comment by hamnis** [comment]
That is one of the unfortunately flaky tests :(

**Comment by rossabaker** [comment]
This looks good, but we've got a merge conflict now.

**Comment by tian000** [comment]
@rossabaker Fixed the merge conflicts!

**Comment by rossabaker** [comment]
Thanks, @tian000. I backported this to series/0.20 as ba7ea8e5c6d516b59378c2f6e17f44280cf967ab, and will merge it forward.

**Pull request closed without merging**